File: java/socketd/src/main/java/org/noear/socketd/core/SessionBase.java
package org.noear.socketd.core;

import java.util.HashMap;
import java.util.Map;

/**
 * 会话基类
 *
 * @author noear
 */
public abstract class SessionBase implements Session {
    protected final Channel channel;

    public SessionBase(Channel channel) {
        this.channel = channel;
    }

    /**
     * 会话的附件与通道的各自独立
     */
    private Map<String, Object> attrMap;

    @Override
    public Map<String, Object> getAttrMap() {
        if (attrMap == null) {
            attrMap = new HashMap<>();
        }

        return attrMap;
    }

    /**
     * 获取附件
     */
    @Override
    public <T> T getAttr(String name) {
        if (attrMap == null) {
            return null;
        }

        return (T) attrMap.get(name);
    }

    /**
     * 设置附件
     */
    @Override
    public <T> void setAttr(String name, T value) {
        if (attrMap == null) {
            attrMap = new HashMap<>();
        }
        attrMap.put(name, value);
    }

    private String sessionId;

    @Override
    public String getSessionId() {
        if (sessionId == null) {
            sessionId = generateKey();
        }

        return sessionId;
    }

    protected String generateKey(){
        return channel.getConfig().getKeyGenerator().generate();
    }
}

File: python/socketd/socketd/transport/core/internal/SessionBase.py
from abc import ABC
from typing import Any, Dict
from socketd.transport.core.Session import Session
from socketd.transport.core.Channel import Channel


class SessionBase(Session, ABC):
    def __init__(self, channel: Channel):
        self._channel = channel
        self._attr_map = None
        self._session_id = None

    def get_attr_map(self) -> Dict[str, Any]:
        if self._attr_map is None:
            self._attr_map = {}

        return self._attr_map

    def get_attr(self, name: str) -> Any:
        if self._attr_map is None:
            return None

        return self._attr_map.get(name)

    def get_attr_or_default(self, name: str, default: Any) -> Any:
        tmp = self.get_attr(name)
        if tmp is None:
            return default
        else:
            return tmp

    def set_attr(self, name: str, value: Any) -> None:
        if self._attr_map is None:
            self._attr_map = {}
        self._attr_map[name] = value

    def get_session_id(self) -> str:
        if self._session_id is None:
            self._session_id = self.generate_id()

        return self._session_id

    def generate_id(self) -> str:
        return self._channel.get_config().get_id_generator()().__str__()

    def set_session_id(self, value):
        self._session_id = value
